src/dmd/dtoh.d

```diff
@@ -0,0 +1,1692 @@
+        transitivevisitor   = Identifier.idPool("transitivevisitor");
+        dmd                 = Identifier.idPool("dmd");
+    }
+}
+
+private struct DMDClass
+{
+    __gshared Identifier ID; ////Identifier
+    __gshared Identifier Visitor;
+    __gshared Identifier ParseTimeVisitor;
+    static void _init()
+    {
+        ID                  = Identifier.idPool("Identifier");
+        Visitor             = Identifier.idPool("Visitor");
+        ParseTimeVisitor    = Identifier.idPool("ParseTimeVisitor");
+    }
+
+}
+
+private bool isIdentifierClass(ASTCodegen.ClassDeclaration cd)
+{
+    return (cd.ident == DMDClass.ID &&
+            cd.parent !is null &&
+            cd.parent.ident == DMDModule.identifier &&
+            cd.parent.parent && cd.parent.parent.ident == DMDModule.dmd &&
+            !cd.parent.parent.parent);
+}
+
+private bool isVisitorClass(ASTCodegen.ClassDeclaration cd)
+{
+    for (auto cdb = cd; cdb; cdb = cdb.baseClass)
+    {
+        if (cdb.ident == DMDClass.Visitor ||
+            cdb.ident == DMDClass.ParseTimeVisitor)
+        return true;
+    }
+    return false;
+}
+
+private bool isIgnoredModule(ASTCodegen.Module m)
+{
+    //if (!m || !m.parent)
+    if (!m)
+        return true;
+
+    // Ignore dmd.root
+    if (m.parent && m.parent.ident == DMDModule.root &&
+        m.parent.parent && m.parent.parent.ident == DMDModule.dmd &&
+        !m.parent.parent.parent)
+    {
+        return true;
+    }
+
+    // Ignore dmd.visitor and derivatives
+    if ((m.ident == DMDModule.visitor ||
+            m.ident == DMDModule.parsetimevisitor ||
+            m.ident == DMDModule.permissivevisitor ||
+            m.ident == DMDModule.strictvisitor ||
+            m.ident == DMDModule.transitivevisitor) &&
+            m.parent && m.parent.ident == DMDModule.dmd &&
+            !m.parent.parent)
+    {
+        return true;
+    }
+    return false;
+}
+
+private bool isFrontendModule(ASTCodegen.Module m)
+{
+    if (!m || !m.parent)
+        return false;
+
+    // Ignore dmd.root
+    if (m.parent.ident == DMDModule.root &&
+        m.parent.parent && m.parent.parent.ident == DMDModule.dmd &&
+        !m.parent.parent.parent)
+    {
+        return false;
+    }
+
+    // Ignore dmd.visitor and derivatives
+    if ((m.ident == DMDModule.visitor ||
+            m.ident == DMDModule.parsetimevisitor ||
+            m.ident == DMDModule.permissivevisitor ||
+            m.ident == DMDModule.strictvisitor ||
+            m.ident == DMDModule.transitivevisitor) &&
+            m.parent && m.parent.ident == DMDModule.dmd &&
+            !m.parent.parent)
+    {
+        return false;
+    }
+    return ((m.parent.ident == DMDModule.dmd && !m.parent.parent) ||
+            (m.parent.parent.ident == DMDModule.dmd && !m.parent.parent.parent));
+}
+
+void genCppHdrFiles(ref Modules ms)
+{
+    DMDType._init();
+    if (isBuildingCompiler)
+    {
+        DMDModule._init();
+        DMDClass._init();
+    }
```

Comment:
You might want to protect these initialization with a singleton, s.t. double initialization doesn't happen when DMD is e.g. used as a library.

Reply:
Done